akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2270,3 +2271,70 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
   override def toString = "StatefulMapConcat"
 
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class AggregateWithBoundary[In, Agg, Out](
+    allocate: () => Agg,
+    aggregate: (Agg, In) => Boolean,
+    harvest: Agg => Out,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)])
+    extends GraphStage[FlowShape[In, Out]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      private[this] var aggregated: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay(s"${this.getClass.getSimpleName}Timer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = {
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregated != null && isReadyOnTimer(aggregated)) harvestAndEmit()
+        }
+      }
+
+      // at onPush, isAvailable(in)=true hasBeenPulled(in)=false, isAvailable(out) could be true or false due to timer triggered emit
+      override def onPush(): Unit = {
+        if (aggregated == null) aggregated = allocate()
```

Comment:
@johanandren I like the one liner using `null` instead of pattern match with `OptionVal`.
`null` should not be too much an eyesore if this is internal API with performance in mind.

Reply:
Not a huge thing no, just something that differs from the recent parts of the Akka code base, so I'd appreciate if you could align. A benefit with option val (just like with Option) is that it is unlikely to compile at all if you use it incorrectly by mistake since the actual type is wrapped.

Reply:
sure, I can certainly change it back to get the align with the style, but `GroupedWeightedWithin` also used `null`
https://github.com/akka/akka/blob/d703a2afe0e0c6dc8f1c0321b0f4652600e2ae99/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala#L1760

The only unsafe part with `null` is the equality test `==` and `!=`, but even that is the intended special case.
Scala already force the var definition to be typed, so complier with not allow it to compare with or be assigned any other types except for `null`

Reply:
@johanandren want to wrap up this conversation. Are you ok with keeping it this way? Given `null` is used in other ops, and I cannot see much extra type safety to trade off with the potential verbosity.

Reply:
Yeah, ok, let's leave it like this.